### PR TITLE
[RTL] Support ENABLE_SYSTEMWIDE_DPH define

### PR DESCRIPTION
**Pull request by SergeGautherie**
## Purpose

To enable it, run "configure.*" with "-DENABLE_SYSTEMWIDE_DPH=1".

JIRA issue: [CORE-12020](https://jira.reactos.org/browse/CORE-12020)

Co-Authored-By: Thomas Faber <thomas.faber@reactos.org>

## Proposed changes

This is the basic part only, which matches
https://reactos.org/wiki/Debugging#Debug_Page_Heap_.28DPH.29

### CODE REVIEW

File: sdk/lib/rtl/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,10 @@ add_definitions(
     -D_NTSYSTEM_
     -D_NTDLLBUILD_)
 
+if(ENABLE_SYSTEMWIDE_DPH)
+    add_definitions(-DENABLE_SYSTEMWIDE_DPH)
+endif()
```
Comment (zefklop) [inline sdk/lib/rtl/CMakeLists.txt:10]: Please use `target_compile_definitions` on rtl

**Comment by zefklop** [comment]
Can we enable this globally for debug build ?

I think an option would be in order
https://cmake.org/cmake/help/latest/command/option.html

`option(ENABLE_SYSTEMWIDE_DPH "Globally enable Debug-Paged-Heap" ${DBG})`

Somewhere in sdk/cmake/config.cmake

File: sdk/lib/rtl/CMakeLists.txt (hunk 1 above)
Comment (SergeGautherie) [reply sdk/lib/rtl/CMakeLists.txt:10]: Done.

**Comment by SergeGautherie** [comment]
> Can we enable this globally for debug build ?

I would support setting up some kind of DPH testing.
Nonetheless, your suggestion, as is, is not well suited for that, as this feature makes the build very slow (to install).

Short term: compile a separate build.
Long term: research a way to make this feature dynamically configurable, maybe a registry key...

File: sdk/lib/rtl/heap.c [hunk 2]
```diff
@@ -1289,6 +1293,13 @@ RtlCreateHeap(ULONG Flags,
         Flags &= HEAP_CREATE_VALID_MASK;
     }
 
+#ifdef ENABLE_SYSTEMWIDE_DPH
+    if (!Addr)
+    {
+        Flags |= HEAP_FLAG_PAGE_ALLOCS;
+    }
+#endif
+
```
Comment (zefklop) [inline sdk/lib/rtl/heap.c:1302]: This should not be needed. Look at RtlpPageHeapCreate implementation.

**Review by zefklop — CHANGES_REQUESTED** [review]
Please explain why force-setting HEAP_FLAG_PAGE_ALLOCS is necessary

File: sdk/lib/rtl/heap.c (hunk 2 above)
Comment (learn-more) [reply sdk/lib/rtl/heap.c:1302]: Just setting `FLG_HEAP_PAGE_ALLOCS` on `NtGlobalFlag` should be enough, no need for all these hacks everywhere?

Comment (zefklop) [reply sdk/lib/rtl/heap.c:1302]: NtGlobalFlag is per application, so it would be tedious to enable it globally.

Comment (learn-more) [reply sdk/lib/rtl/heap.c:1302]: There is also a global `NtGlobalFlag`, that all applications inherit

Comment (zefklop) [reply sdk/lib/rtl/heap.c:1302]: Right, it's in Session Manager\GlobalFlag Somehow I missed it.

@SergeGautherie maybe setting it in regtest.cmd would do what you want to achieve.

**Comment by SergeGautherie** [comment]
>  this feature makes ReactOS very slow (to install)

Let's wait for #3517 first...

**Comment by zefklop** [comment]
> > this feature makes ReactOS very slow (to install)
> 
> Let's wait for #3517 first...

This won't change anything, DPH is not using the code paths that #3517 solves

File: sdk/lib/rtl/heap.c (hunk 2 above)
Comment (SergeGautherie) [reply sdk/lib/rtl/heap.c:1302]: > Just setting `FLG_HEAP_PAGE_ALLOCS` on `NtGlobalFlag` should be enough, no need for all these hacks everywhere?

Thanks for the hint.
I tried to do it on NTOS:EX and NTDLL:LDR, but that did not succeed.
Could you tell me where/how to do that exactly?

--

> regtest.cmd

That might be a good idea for running tests, but it is not the system-wide case (from ISO start).

**Review by SergeGautherie — COMMENTED** [review]
> Please explain why force-setting HEAP_FLAG_PAGE_ALLOCS is necessary

Before looking into that later, I would simply remind that I just copied what the wiki suggests since 8 years ago.

**Comment by GeoB99** [comment]
Stale PR.

**Pull request closed without merging**